src/main/java/httpServer/ListenerAndResponder.java

```diff
@@ -40,46 +40,102 @@ public Socket keepListening() throws IOException {
+
+        HTTPRequestBuilder requestBuilder = new HTTPRequestBuilder();
+        requestBuilder = readHeaders(bufferedReader, requestBuilder);
+        int bodyLength = requestBuilder.getContentLengthInt();
+        StringBuilder body = readBody(bufferedReader, bodyLength);
+        return returnHttpRequest(body, requestBuilder);
 
     }
 
-    private String readAllLines(ClientReadable bufferedReader)  {
-        String line;
-        StringBuilder stringBuilder = new StringBuilder();
+    private HTTPRequestBuilder readHeaders(ClientReadable bufferedReader, HTTPRequestBuilder requestBuilder) {
+
+        String character;
+        StringBuilder headers = new StringBuilder();
+
         try {
-            while ((line = bufferedReader.readLine()) != null)
-            {
-                stringBuilder.append(line);
-                stringBuilder.append(CRLF);
-                if (line.equals("")) { break; }
+            while ((character = bufferedReader.read()) != null) {
+
+                headers.append(character);
+                if (checkIfReachedEndOfHeaders(headers)) {
+                    break;
+                }
             }
-            return stringBuilder.toString();
+
         } catch (IOException e) {
             e.printStackTrace();
-            return stringBuilder.toString();
         }
+        String[] headersArray = headers.toString().split(CRLF);
+        for (String header: headersArray) {
+            requestBuilder = discernHeaderTypeAndAddToRequestBuilder(header, requestBuilder);
+        }
+        return requestBuilder;
+
     }
+
+    private StringBuilder readBody(ClientReadable bufferedReader, int bodyLength) {
+        String character;
+        StringBuilder body = new StringBuilder();
+        if (bodyLength == 0) {
+            return body;
+        }
+        try {
+            while ((character = bufferedReader.read()) != null) {
+                body.append(character);
+                if (body.length() == bodyLength) {
+                    break;
+                }
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return body;
+    }
+
+    private HTTPRequest returnHttpRequest(StringBuilder body, HTTPRequestBuilder requestBuilder) {
+        if (body.length() != 0) {
+            requestBuilder.setBody(body.toString());
+        }
+        return requestBuilder.build();
+    }
+
+    private HTTPRequestBuilder discernHeaderTypeAndAddToRequestBuilder(String line, HTTPRequestBuilder httpRequestBuilder) {
+        if (line.contains(": ")) {
+            httpRequestBuilder.buildHeaderLine(line);
+        } else {
+            httpRequestBuilder.buildRequestLine(line);
+        }
+        return httpRequestBuilder;
+    }
+
+    private boolean checkIfReachedEndOfHeaders(StringBuilder requestScannedInSoFar) {
+        return requestScannedInSoFar.toString().contains(CRLF + CRLF);
+    }
```

Comment:
There's overall a _lot_ of private methods relating to the reading in of the request that feels like maybe it should be in its own class. I won't make it a condition for merge of this PR, but I will ask you to do some extraction for your next PR.